lib/sgtn-client/api/translation.rb

```diff
@@ -137,12 +124,84 @@ def self.load_s(component, locale)
         rescue => exception
           SgtnClient.logger.error exception.message
         end
-        if obj != nil
-          obj = obj["data"]
+        obj && obj["data"]
+      end
+
+    def self.load_and_compare_source(component, locale)
+      source_cache_key = SgtnClient::CacheUtil.get_cachekey(component, LocaleUtil.get_source_locale)
+      _, source_bundle = SgtnClient::CacheUtil.get_cache(source_cache_key)
+      if source_bundle.nil? || source_bundle.empty?
+        source_bundle = single_load(component, SgtnClient::Config.configurations.default).value
+        SgtnClient::CacheUtil.write_cache(source_cache_key, source_bundle)
+      end
+      return source_bundle if LocaleUtil.is_source_locale(locale)
+
+      translation_bundle_thread = single_load(component, locale)
+      old_source_bundle_thread = single_load(component, LocaleUtil.get_source_locale)
+      translation_bundle = translation_bundle_thread.value
+      old_source_bundle = old_source_bundle_thread.value
+
+      if translation_bundle.nil? || source_bundle.nil? || old_source_bundle.nil? ||
+         translation_bundle.empty? || source_bundle.empty? || old_source_bundle.empty?
+        return translation_bundle
+      end
+
+      source_messages = source_bundle.first[1]
+      old_source_messages = old_source_bundle['messages']
+      translation_messages = translation_bundle['messages']
+      new_translation_messages = {}
+      source_messages.each do |key, value|
+        new_translation_messages[key] = if old_source_messages[key] == value
+                                          translation_messages[key] || value
+                                        else
+                                          value
+                                        end
+      end
+      translation_bundle['messages'] = new_translation_messages
+      translation_bundle
+    end
+
+    @@load_threads_lock = Mutex.new
+    @@load_threads = {}
+    def self.single_load(component, locale)
+      cache_key = SgtnClient::CacheUtil.get_cachekey(component, locale)
+      @@load_threads_lock.synchronize do
+        thread = @@load_threads[cache_key]
+        if thread.nil? || thread.alive? == false
+          thread = Thread.new do
+            if locale == SgtnClient::Config.configurations.default
+              Source.getBundle(component, SgtnClient::Config.configurations.default)
+            else
+              load(component, locale)
+            end
+          end
+          @@load_threads[cache_key] = thread
+        end
+        return thread
+      end
+    end
+
+    @@refresh_threads_lock = Mutex.new
+    @@refresh_threads = {}
+    def self.single_refresh(component, locale)
+      cache_key = SgtnClient::CacheUtil.get_cachekey(component, locale)
+      @@refresh_threads_lock.synchronize do
```

Comment:
Only one thread in entire app can access this code block at a time. For better performance, it should be on1 thread per cached bundle.

Reply:
This is necessary because `@@refresh_threads` is read at the beginning of this block and is assigned at the end.
No one statement can be moved out of this block.
This block is fast actually because it only starts a thread.

> For better performance, it should be 1 thread per cached bundle.
This is true bacause `@@refresh_threads` is a hash, each cached budle has an entry in it.